java-viewer/src/main/java/org/sonar/java/viewer/DotHelper.java

```diff
@@ -0,0 +1,119 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.viewer;
+
+import org.sonar.plugins.java.api.tree.Tree;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+import java.text.MessageFormat;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DotHelper {
+
+  private DotHelper() {
+  }
+
+  public static String node(int id, String label, @Nullable Highlighting highlighting) {
+    return node(id, label, highlighting, Collections.emptyMap());
+  }
+
+  public static String node(int id, String label, @Nullable Highlighting highlighting, Map<String, String> extraFields) {
+    Map<String, String> newMap = new HashMap<>(extraFields);
+    newMap.put("label", label);
+    if (highlighting != null) {
+      newMap.put("highlighting", highlighting.name);
+    }
+    return MessageFormat.format("{0}[{1}];", id, extraFields(newMap));
+  }
+
+  public static String edge(int from, int to, @Nullable String label) {
+    return edge(from, to, label, null, Collections.emptyMap());
+  }
+
+  public static String edge(int from, int to, @Nullable String label, Highlighting highlighting) {
+    return edge(from, to, label, highlighting, Collections.emptyMap());
+  }
+
+  public static String edge(int from, int to, @Nullable String label, @Nullable Highlighting highlighting, Map<String, String> extraFields) {
+    Map<String, String> newMap = new HashMap<>(extraFields);
+    if (label != null) {
+      newMap.put("label", label);
+    }
+    if (highlighting != null) {
+      newMap.put("highlighting", highlighting.name);
+    }
+    return MessageFormat.format("{0}->{1}[{2}];", from, to, extraFields(newMap));
+  }
+
+  private static String extraFields(Map<String, String> extraFields) {
+    if (extraFields.isEmpty()) {
+      return "";
+    }
+    return extraFields.entrySet().stream()
+      .filter(entry -> entry.getValue() != null)
+      .map(entry -> MessageFormat.format("{0}=\"{1}\"", entry.getKey(), entry.getValue()))
+      .collect(Collectors.joining(","));
+  }
+
+  public enum Highlighting {
+    FIRST_NODE("firstNode"),
+    LOST_NODE("lostNode"),
+    EXIT_NODE("exitNode"),
+
+    TOKEN_KIND("tokenKind"),
+    CLASS_KIND("classKind"),
+    METHOD_KIND("methodKind"),
+
+    EXCEPTION_EDGE("exceptionEdge"),
+    YIELD_EDGE("yieldEdge");
+
+    private final String name;
+
+    private Highlighting(String name) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') "private" is redundant in this context. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2333)